### PR TITLE
[Discuss] cudf support.

### PR DESCRIPTION
**Pull request by trivialfis**
@RAMitchell This is early PR for discussion.  I will keep pushing changes to my own private branch instead of this one.   How do we wanna proceed?

To run a small test see if it crashes :-)
``` python
import numpy as np
import pandas as pd
import cudf
import xgboost as xgb

kRows = 10
kCols = 2

na = np.random.randn(kRows, kCols).astype(np.float32)
na[3, 1] = np.NAN
pa = pd.DataFrame(na)

names = []

for i in range(0, kCols):
    names.append(str(i))
pa.columns = names

cd: cudf.DataFrame = cudf.from_pandas(pa)

dtrain = xgb.DMatrix(cd)
```

### CODE REVIEW

**Comment by trivialfis** [comment]
@mt-jones @canonizer

File: src/data/simple_csr_source.cu [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+  info.num_col_     = n_cols;
+  info.num_row_     = cols[0].size;
+  info.num_nonzero_ = n_valid;
+
+  GPUSet devices = GPUSet::Range(0, 1);
+
+  page_.offset.Reshard(GPUDistribution(devices));
+  page_.offset.Resize(info.num_row_ + 1);
+  // page_.offset.Fill(1);
```
Comment (trivialfis) [inline src/data/simple_csr_source.cu]: Somehow calling Fill results in a failed span check in later kernel.

File: src/data/data.cu [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+#include "xgboost/data.h"
+#include "xgboost/logging.h"
+#include "xgboost/json.h"
+
+#include "columnar.h"
+#include "../common/device_helpers.cuh"
+
+namespace xgboost {
+
+void MetaInfo::SetInfo(const char * c_key, std::string const& interface_str) {
```
Comment (RAMitchell) [inline src/data/data.cu]: This interface string is always from cuda array interface right? I think you should name variables/functions implying that it is device memory.

Comment (trivialfis) [reply src/data/data.cu]: @RAMitchell For now yes.  Here's a question I need to answer myself:

Interface is not designed to distinguish where is the data pointer.  So I have two options:

- Ask user to provide a `context` field in each interface object.
- Split it into two functions for host side array and device side.

Which one should I prefer.

**Comment by trivialfis** [comment]
- How would we modify this code to construct a DMatrix given a 2d cupy array?

I think it's better to make separated code path.

- Is the page construction on the device sparsity aware? i.e. does it allocate memory proportional only to non-zeros

Yes,  `Resize(n_valid)`.

- What are your thoughts on also allowing construction of the SparsePage on the host?

We do both.

File: src/data/data.cu (hunk 2 above)
Comment (RAMitchell) [reply src/data/data.cu]: Doesn't cuda array interface imply the pointer always points to device memory? We are not trying to support numpys array interface here.

Note that you can always check if a pointer resides on host or device using cudaPointerGetAttributes.

Comment (trivialfis) [reply src/data/data.cu]: Host memory is on my to-do list, Apache Arrow is such a big family, but let's talk about it on another day.

**Comment by trivialfis** [comment]
One important issue is, how do we assign the correct gpu-id for training?  Do we prioritize the global variable mentioned before?  I might need a lots of help regarding language bindings.

**Pull request closed without merging**